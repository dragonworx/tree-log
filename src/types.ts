export type Arguments = any[];

export interface LogDetail {
  timestamp: Date;
  label: string;
}

export interface LogNode extends LogDetail {
  parent?: LogNode;
  children: Array<LogItem>;
}

export interface LogEntry extends LogDetail {
  data?: Arguments;
  children?: LogEntry[];
}

export interface LogTrace extends LogEntry {
  depth?: number;
  isLastChild?: boolean;
  isNode?: boolean;
}

export interface LogTraceDetail {
  depth: number;
  isLastChild: boolean;
  isNode: boolean;
}

export type LogItem = LogNode | LogEntry;

export interface State {
  options: LogOptions;
  root: LogNode;
  head: LogNode;
}

export interface LogOptions {
  enabled: boolean;
  showTimeStamp: boolean;
  useTimeDelta: boolean;
  useColor: boolean;
  stringProviderMethodName: string;
}
