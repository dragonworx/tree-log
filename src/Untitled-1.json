{
  "id": 0,
  "timestamp": "2022-05-14T10:01:53.060Z",
  "identifier": "__ROOT__",
  "children": [
    {
      "id": 1,
      "timestamp": "2022-05-14T10:01:53.061Z",
      "identifier": "foo",
      "args": [],
      "children": [
        {
          "id": 2,
          "timestamp": "2022-05-14T10:01:53.061Z",
          "identifier": "bar",
          "args": [1],
          "children": [
            {
              "id": 3,
              "timestamp": "2022-05-14T10:01:53.061Z",
              "identifier": "baz",
              "args": [1, {}, "2022-05-14T10:01:53.061Z"],
              "children": [
                {
                  "id": 4,
                  "timestamp": "2022-05-14T10:01:53.061Z",
                  "identifier": "bez",
                  "args": [],
                  "children": [
                    {
                      "id": 5,
                      "timestamp": "2022-05-14T10:01:53.061Z",
                      "identifier": "efg",
                      "args": [3, false, ["a", "b", true]],
                      "children": []
                    },
                    {
                      "id": 6,
                      "timestamp": "2022-05-14T10:01:53.061Z",
                      "identifier": "abc",
                      "args": [3, true],
                      "children": []
                    }
                  ]
                },
                {
                  "id": 7,
                  "timestamp": "2022-05-14T10:01:53.062Z",
                  "identifier": "xyz",
                  "args": [
                    4,
                    {
                      "x": 1
                    }
                  ],
                  "children": []
                },
                {
                  "id": 8,
                  "timestamp": "2022-05-14T10:01:53.062Z",
                  "identifier": "xyz",
                  "args": [4],
                  "children": [
                    {
                      "id": 9,
                      "timestamp": "2022-05-14T10:01:53.062Z",
                      "identifier": "xyz",
                      "args": [5, null, null],
                      "children": [
                        {
                          "id": 10,
                          "timestamp": "2022-05-14T10:01:53.062Z",
                          "identifier": "xyz",
                          "args": [6, {}],
                          "children": []
                        },
                        {
                          "id": 11,
                          "timestamp": "2022-05-14T10:01:53.062Z",
                          "identifier": "xyz",
                          "args": [6],
                          "children": []
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
