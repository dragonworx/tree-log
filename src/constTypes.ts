export type Arguments = any[];

export const isBrowser = typeof window !== 'undefined';

export const RootLabel = '__root__';

export const defaultLogOptions: LogOptions = {
  enabled: true,
  showTimeStamp: true,
  useTimeDelta: false,
  useColor: true,
  stringProviderMethodName: 'toLogInfo',
};

export const state: State = (() => {
  const root: LogNode = {
    timestamp: new Date(),
    label: RootLabel,
    children: [],
  };

  return {
    hasWarned: false,
    options: defaultLogOptions,
    root,
    head: root,
  };
})();

export interface LogDetail {
  timestamp: Date;
  label: string;
}

export interface LogNode extends LogDetail {
  parent?: LogNode;
  children: Array<LogItem>;
}

export interface LogEntry extends LogDetail {
  data?: Arguments;
  children?: LogEntry[];
}

export interface LogTrace extends LogEntry {
  depth?: number;
  isLastChild?: boolean;
  isNode?: boolean;
}

export interface LogTraceDetail {
  depth: number;
  isLastChild: boolean;
  isNode: boolean;
}

export type LogItem = LogNode | LogEntry;

export interface State {
  hasWarned: boolean;
  options: LogOptions;
  root: LogNode;
  head: LogNode;
}

export interface LogOptions {
  enabled: boolean;
  showTimeStamp: boolean;
  useTimeDelta: boolean;
  useColor: boolean;
  stringProviderMethodName: string;
}

export interface ColorOptions {
  bold?: boolean;
  underline?: boolean;
}
