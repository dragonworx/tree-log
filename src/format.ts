import { defaultStringifyOptions } from ".";
import { Arguments, LogDetail, LogEntry, StringifyOptions } from "./types";
const c = require("ansi-colors");

// https://en.wikipedia.org/wiki/Alt_code

export function dataToString(
  data: Arguments,
  options: StringifyOptions
): string {
  const { stringProviderMethodName } = {
    ...defaultStringifyOptions,
    ...options,
  };

  return data
    .map((value) => {
      const type = typeof value;
      if (type === "string") {
        return c.white(`"${value}"`);
      } else if (type === "number") {
        return c.green(value);
      } else if (type === "boolean") {
        return c.magenta(value);
      } else if (value === null) {
        return c.red("null");
      } else if (value === undefined) {
        return c.red("undefined");
      } else if (type === "object") {
        const proto = value.__proto__.constructor.name;
        if (proto === "Array") {
          return "[" + dataToString(value, options) + "]";
        } else if (proto === "Date") {
          return c.cyanBright(`${formatDate(value)} ${formatTime(value)}`);
        } else if (proto === "RegExp") {
          const regex = value as RegExp;
          return c.cyanBright(`/${regex.source}/${regex.flags}`);
        } else {
          return c.cyanBright(
            stringProviderMethodName! in value
              ? dataToString([value[stringProviderMethodName!]()], options)
              : "{ " +
                  Object.keys(value)
                    .map(
                      (key) => `${key}: ${dataToString([value[key]], options)}`
                    )
                    .join(", ") +
                  " }"
          );
        }
      }
      return value;
    })
    .join(c.grey(", "));
}

export function entryToString(
  entry: LogEntry,
  previousDetail: LogDetail,
  depth: number,
  options: StringifyOptions = {}
) {
  const { showTimestamp, isLastChild, useColor, useTimeDelta } = {
    ...defaultStringifyOptions,
    ...options,
  };

  c.enabled = !!useColor;

  const { identifier, data } = entry;

  const prefix = showTimestamp
    ? `${formatTimeStamp(entry, previousDetail, useTimeDelta!)} `
    : "";
  const indent = c.grey(`│ `.repeat(depth) + (isLastChild ? "└" : "├"));
  const id = c.bold.yellow(`${identifier} `);
  const dataStr = data ? dataToString(data, options) : "";

  return `${prefix}${indent}${id}${dataStr}`;
}

export function formatTimeStamp(
  entry: LogEntry,
  previousDetail: LogDetail,
  useTimeDelta: boolean
) {
  const { timestamp } = entry;

  if (useTimeDelta) {
    const previousTimestamp = previousDetail.timestamp;
    const delta = timestamp.getTime() - previousTimestamp.getTime();

    return `+${c.cyanBright(delta)}`;
  } else {
    const dateStr = formatDate(timestamp);
    const timeStr = formatTime(timestamp);

    return `${c.blue(dateStr + "│")}${c.cyanBright(timeStr + "│")}`;
  }
}

export function formatDate(date: Date) {
  return `${String(date.getMonth() + 1).padStart(
    2,
    "0"
  )}/${date.getDate()}/${date.getFullYear()}`;
}

export function formatTime(date: Date) {
  const hours = date.getHours();
  const mins = date.getMinutes();
  const secs = date.getSeconds();
  const ms = date.getMilliseconds();
  const minutes = mins < 10 ? "0" + mins : mins;

  return `${hours}:${minutes}:${secs}:${ms}`;
}
