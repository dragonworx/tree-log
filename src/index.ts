export {
  log,
  pushLog,
  popLog,
  clearLog,
  renderLog,
  setLogOptions,
} from './log';
