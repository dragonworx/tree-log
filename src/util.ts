import { defaultStringifyOptions } from "./";
import { Arguments, LogEntry, StringifyOptions } from "./types";
const c = require("ansi-colors");

// https://en.wikipedia.org/wiki/Alt_code

export function dataToString(
  data: Arguments,
  options: StringifyOptions
): string {
  const { useColor, stringProviderMethodName } = {
    ...defaultStringifyOptions,
    ...options,
  };

  c.enabled = !!useColor;

  return data
    .map((value) => {
      const type = typeof value;
      if (type === "string") {
        return `"${c.yellow(value)}"`;
      } else if (type === "number") {
        return c.green(value);
      } else if (type === "boolean") {
        return c.cyanBright(value);
      } else if (value === null) {
        return c.grey("null");
      } else if (value === undefined) {
        return c.red("undefined");
      } else if (type === "object") {
        const proto = value.__proto__.constructor.name;
        if (proto === "Array") {
          return "[" + dataToString(value, options) + "]";
        } else if (proto === "Date") {
          return `${formatDate(value)} ${formatTime(value)}`;
        } else if (proto === "RegExp") {
          const regex = value as RegExp;
          return c.yellow(`/${regex.source}/${regex.flags}`);
        } else {
          return stringProviderMethodName! in value
            ? dataToString([value[stringProviderMethodName!]()], options)
            : "{ " +
                Object.keys(value)
                  .map(
                    (key) => `${key}: ${dataToString([value[key]], options)}`
                  )
                  .join(", ") +
                " }";
        }
      }
      return value;
    })
    .join(c.grey(", "));
}

export function entryToString(
  entry: LogEntry,
  depth: number,
  options: StringifyOptions = {}
) {
  const { showTimestamp } = {
    ...defaultStringifyOptions,
    ...options,
  };

  const { timestamp, identifier, args } = entry;

  const prefix = showTimestamp
    ? `${formatTimeStamp(timestamp)}${c.bold.blueBright("│")} `
    : "";
  const indent = c.grey(
    `│ `.repeat(Math.max(0, depth - 1)) + entry.parent?.identifier
  );
  const id = c.bold.white(`${identifier}: `);
  const dataStr = args ? dataToString(args, options) : "";

  return `${prefix}${indent}${id}${dataStr}`;
}

export function formatTimeStamp(date: Date) {
  const dateStr = formatDate(date);
  const timeStr = formatTime(date);

  return `${c.blue(dateStr)} ${c.cyanBright(timeStr)}`;
}

export function formatDate(date: Date) {
  return `${date.getMonth() + 1}/${date.getDate()}/${date.getFullYear()}`;
}

export function formatTime(date: Date) {
  const hours = date.getHours();
  const mins = date.getMinutes();
  const secs = date.getSeconds();
  const ms = date.getMilliseconds();
  const minutes = mins < 10 ? "0" + mins : mins;

  return `${hours}:${minutes}:${secs}:${ms}`;
}
