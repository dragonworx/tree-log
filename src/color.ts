import { state, isBrowser, ColorOptions } from './constTypes';

export const ops: Record<string, string> = {
  reset: '\u001b[0m',
  bold: '\u001b[1m',
  underline: '\u001b[4m',
  reversed: '\u001b[7m',
};

export const colors: Record<string, string[]> = {
  black: ['\u001b[30m', 'black'],
  red: ['\u001b[31m', 'red'],
  green: ['\u001b[32m', 'green'],
  yellow: ['\u001b[33m', 'yellow'],
  blue: ['\u001b[34m', '#1950b4'],
  magenta: ['\u001b[35m', 'magenta'],
  cyan: ['\u001b[36m', 'cyan'],
  white: ['\u001b[37m', 'white'],
  //
  grey: ['\u001b[38;5;240m', 'grey'],
  orange: ['\u001b[38;5;220m', 'orange'],
};

export const boldTerminalColors: Record<string, string> = {
  black: '\u001b[30;1m',
  red: '\u001b[31;1m',
  green: '\u001b[32;1m',
  yellow: '\u001b[33;1m',
  blue: '\u001b[34;1m',
  magenta: '\u001b[35;1m',
  cyan: '\u001b[36;1m',
  white: '\u001b[37;1m',
};

const t1 = '\u0001';
const t2 = '\u0002';
const t3 = '\u0003';
const t4 = '\u0004';

export function color(str: string, cname: string, options: ColorOptions = {}) {
  if (!state.options.useColor) {
    return str;
  }
  return `${t2}${cname}+${options.bold ? 1 : 0}+${
    options.underline ? 1 : 0
  }${t3}${str}`;
}

export function decodeColorForLine(str: string) {
  if (!state.options.useColor) {
    return str;
  }
  return str
    .split(t2)
    .map((chunk) => {
      const [cnameInfo, ...contents] = chunk.split('\u0003');
      const str = contents.join('');
      return !isBrowser
        ? terminalColor(str, cnameInfo)
        : browserColor(str, cnameInfo);
    })
    .join('');
}

export function terminalColor(str: string, cnameInfo: string) {
  const [cname, boldFlag, underlineFlag] = cnameInfo.split('+');
  if (cname.length === 0) {
    return str;
  }
  const bold = boldFlag === '1';
  const underline = underlineFlag === '1';
  const setColor = bold ? `${boldTerminalColors[cname]}` : colors[cname][0];
  return `${setColor}${underline ? ops.underline : ''}${str}${ops.reset}`;
}

export function browserColor(str: string, cnameInfo: string) {
  return `${t1}${cnameInfo}${t4}${str}`;
}

export function decodeBrowserColorLine(str: string): [string, string[]] {
  const style: string[] = [];
  const output = str
    .split(t1)
    .map((chunk) => {
      const [cnameInfo, ...message] = chunk.split(t4);
      const [cname, boldFlag, underlineFlag] = cnameInfo.split('+');
      if (cname.length === 0) {
        return message;
      }
      const bold = boldFlag === '1';
      const underline = underlineFlag === '1';
      const colorStyle = `color:${colors[cname][1]};`;
      const boldStyle = bold ? 'text-weight:bold;' : '';
      const underlineStyle = underline ? 'text-decoration:underline;' : '';
      style.push(
        colorStyle + boldStyle + underlineStyle + 'font-family:monospace;',
      );
      return `%c${message}`;
    })
    .join('');
  return [output, style];
}
