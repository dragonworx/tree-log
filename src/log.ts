import { LogNode, LogEntry, State, LogOptions } from './types';
import { entryToString } from './format';
import { flatten } from './write';

const RootLabel = '__root__';

export const defaultLogOptions: LogOptions = {
  enabled: false,
  showTimeStamp: true,
  useTimeDelta: false,
  useColor: true,
  stringProviderMethodName: 'toLogInfo',
};

export const state: State = (() => {
  const root: LogNode = {
    timestamp: new Date(),
    label: RootLabel,
    children: [],
  };

  return {
    options: defaultLogOptions,
    root,
    head: root,
  };
})();

export function log(label: string, ...data: any[]) {
  if (!state.options.enabled) {
    return;
  }

  const entry: LogEntry = {
    timestamp: new Date(),
    label,
    data,
  };

  state.head.children.push(entry);
}

export function pushLog(label: string) {
  if (!state.options.enabled) {
    return;
  }

  const node: LogNode = {
    timestamp: new Date(),
    label,
    parent: state.head,
    children: [],
  };

  state.head.children.push(node);
  state.head = node;
}

export function popLog() {
  if (!state.options.enabled) {
    return;
  }

  if (state.head.parent) {
    state.head = state.head.parent;
  }
}

export function renderLog() {
  const flat = flatten(true);
  const buffer: Array<string> = [];

  flat.forEach((entry, i) => {
    buffer.push(
      entryToString(
        entry,
        i > 0 ? flat[i - 1] : entry,
        {
          depth: entry.depth!,
          isLastChild: entry.isLastChild!,
          isNode: entry.isNode!,
        },
        state.options,
        entry.isLastChild,
      ),
    );
  });

  return buffer.join('\n');
}

export function setLogOptions(options: Partial<LogOptions>) {
  state.options = {
    ...state.options,
    ...options,
  };
}

export function clearLog() {
  state.root = state.head = {
    timestamp: new Date(),
    label: RootLabel,
    children: [],
  };
}
