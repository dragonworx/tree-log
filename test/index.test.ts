import {
  setEnabled,
  clear,
  log,
  push,
  pop,
  render,
  toArray,
  flatten,
  root,
} from "../src";
import { LogEntry } from "../src/types";

setEnabled(true);

describe("Tree Log", () => {
  beforeEach(() => {
    clear();
  });

  describe("Logging", () => {
    const args = [1, "two", true];

    const setup = () => {
      log("some-identifier", ...args);
      const children = root().children;
      const entry = children[0] as LogEntry;
      return { children, entry };
    };

    it("should capture identifier", () => {
      const { children, entry } = setup();
      expect(children).toHaveLength(1);
      expect(entry.identifier).toBe("some-identifier");
    });

    it("should capture data", () => {
      const { entry } = setup();
      expect(entry.data).toEqual(args);
    });

    it("should capture timestamp", () => {
      const { entry } = setup();
      expect(entry.timestamp).toBeInstanceOf(Date);
    });
  });

  it("should match snapshot", (done) => {
    log("1");
    push("2");
    log("2.1", 1, /a/g, new Date(1659007228539));
    push("3");
    log("3.1", 3, false, ["a", "b", true]);
    log("3.2", 3, true);
    pop();
    log("2.2", 4, { x: 1 });
    pop();

    setTimeout(() => {
      log("1.1", 4, null, undefined);
      push("2");
      log("2.1", 6, {
        toLogInfo() {
          return { x: 1, y: 2, w: { z: [1, 2, true] } };
        },
      });

      setTimeout(() => {
        pop();
        log("1.3", 6);

        const outputSnapshot = render({
          useColor: false,
          showTimestamp: false,
        });

        const outputFull = render({
          useColor: false,
          showTimestamp: false,
        });

        console.log(outputFull);
        console.log(JSON.stringify(toArray(), null, 4));
        console.log(JSON.stringify(flatten(false), null, 4));
        expect(outputSnapshot).toMatchSnapshot();
        done();
      }, 1000);
    }, 500);
  });
});
