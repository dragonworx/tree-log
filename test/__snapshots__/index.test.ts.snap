// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Log should match snapshot for output 1`] = `
"├─1 
│ ├─2 
│ ├─2.1 1, /a/g, 01/1/1970 0:00:0:0
│ │ ├─3 
│ │ ├─3.1 3, false, [\\"a\\", \\"b\\", true]
│ │ └─3.2 3, true
│ └─2.2 4, { x: 1 }
├─1.1 4, null, undefined
│ ├─2 
│ └─2.1 6, { x: 1, y: 2, w: { z: [1, 2, true] } }
└─1.3 6"
`;
